code/hsec-tools/src/Security/Advisories/Parse.hs

```diff
@@ -1,595 +1,187 @@
-
-instance Toml.ToTable Affected where
-  toTable x = Toml.table $
-    [ "package" Toml..= affectedPackage x
-    , "cvss"    Toml..= affectedCVSS x
-    , "versions" Toml..= affectedVersions x
-    ] ++
-    [ "os"   Toml..= y | Just y <- [affectedOS x]] ++
-    [ "arch" Toml..= y | Just y <- [affectedArchitectures x]] ++
-    [ "declarations" Toml..= asTable (affectedDeclarations x) | not (null (affectedDeclarations x))]
-    where
-      asTable kvs = Map.fromList [(T.unpack k, v) | (k,v) <- kvs]
-
-instance Toml.FromValue AffectedVersionRange where
-  fromValue = Toml.parseTableFromValue $
-   do introduced <- Toml.reqKey "introduced"
-      fixed      <- Toml.optKey "fixed"
-      pure AffectedVersionRange {
-        affectedVersionRangeIntroduced = introduced,
-        affectedVersionRangeFixed = fixed
-        }
-
-instance Toml.ToValue AffectedVersionRange where
-  toValue = Toml.defaultTableToValue
-
-instance Toml.ToTable AffectedVersionRange where
-  toTable x = Toml.table $
-    ("introduced" Toml..= affectedVersionRangeIntroduced x) :
-    ["fixed" Toml..= y | Just y <- [affectedVersionRangeFixed x]]
-
-
-instance Toml.FromValue HsecId where
-  fromValue v =
-   do s <- Toml.fromValue v
-      case parseHsecId s of
-        Nothing -> Toml.failAt (Toml.valueAnn v) "invalid HSEC-ID: expected HSEC-[0-9]{4,}-[0-9]{4,}"
-        Just x -> pure x
-
-instance Toml.ToValue HsecId where
-  toValue = Toml.toValue . printHsecId
-
-instance Toml.FromValue CAPEC where
-  fromValue v = CAPEC <$> Toml.fromValue v
-
-instance Toml.ToValue CAPEC where
-  toValue (CAPEC x) = Toml.toValue x
-
-instance Toml.FromValue CWE where
-  fromValue v = CWE <$> Toml.fromValue v
-
-instance Toml.ToValue CWE where
-  toValue (CWE x) = Toml.toValue x
-
-instance Toml.FromValue Keyword where
-  fromValue v = Keyword <$> Toml.fromValue v
-
-instance Toml.ToValue Keyword where
-  toValue (Keyword x) = Toml.toValue x
-
--- | Get a datetime with the timezone defaulted to UTC and the time defaulted to midnight
-getDefaultedZonedTime :: Toml.Value' l -> Toml.Matcher l ZonedTime
-getDefaultedZonedTime (Toml.ZonedTime' _ x) = pure x
-getDefaultedZonedTime (Toml.LocalTime' _ x) = pure (ZonedTime x utc)
-getDefaultedZonedTime (Toml.Day' _       x) = pure (ZonedTime (LocalTime x midnight) utc)
-getDefaultedZonedTime v                     = Toml.failAt (Toml.valueAnn v) "expected a date with optional time and timezone"
+    -- case parseAdvisoryToml attrs policy doc summary details html toml of
+    --   Left es -> Left (AdvisoryError es (T.pack (unlines (map undefined es))))
+    --   Right adv -> pure adv
+
+    firstPretty ::
+        (e -> T.Text -> ParseAdvisoryError) ->
+        (e -> T.Text) ->
+        Either e a ->
+        Either ParseAdvisoryError a
+    firstPretty ctr f = first $ mkPretty ctr f
+
+    mkPretty ::
+        (e -> T.Text -> ParseAdvisoryError) ->
+        (e -> T.Text) ->
+        e ->
+        ParseAdvisoryError
+    mkPretty ctr f x = ctr x $ f x
```

Comment:
It is usually nicer to, additionally to a `import Data.Text qualified as T` add an import `import Data.Text (Text)`

Reply:
I can change it but it was in the original code

Reply:
Aha! Sure :)